### PR TITLE
feat(usage): include Grok Build in the usage view

### PR DESCRIPTION
**Pull request by EliasofIX**
## Summary

The usage page only scanned Claude Code and Codex session transcripts, so Grok Build work never showed up even though sessions under `~/.grok` already carry per-turn token and cost data.

This adds Grok as a third usage provider end to end:

- Parse `turn_completed` usage from Grok `updates.jsonl` (per-model `modelUsage`, cache breakdown, `costUsdTicks` at 10^10 ticks per USD)
- Scan `$GROK_HOME/sessions` (default `~/.grok/sessions`), restricted to `updates.jsonl` so large chat/events logs stay out of the cold path
- Surface **Grok Build** in the web and mobile usage charts/labels
- Bump the usage contract version for the new provider kind

## Test plan

- [ ] Open Usage after pairing a local server that has Grok sessions in `~/.grok`
- [ ] Confirm **Grok Build** appears in the provider split, daily chart, and model table
- [ ] Confirm Claude/Codex totals still match the previous behavior
- [ ] `vp test run apps/server/src/usage/usageTranscripts.test.ts`

Made with Grok Build (xAI).

---

> [!NOTE]
> **Medium Risk**
> New transcript parsing and cost pro-rating logic could mis-attribute tokens or dollars for edge-case Grok logs, but changes are confined to usage scanning/aggregation with broad test coverage and a contract version bump for older clients.
> 
> **Overview**
> **Grok Build** is added as a third usage provider alongside Claude Code and Codex, so local Grok sessions under `~/.grok` (or `GROK_HOME`) show up on the usage page.
> 
> The server scans **`~/.grok/sessions/**/updates.jsonl`** only—skipping large `chat_history` / `events` logs—and parses **`turn_completed`** lines into per-model usage records with cache breakdowns and **`costUsdTicks`** converted to USD (10¹⁰ ticks per dollar). Multi-model turns get **pro-rated cost** when per-model ticks are missing or only partially reported. The usage contract adds **`grok`** to `UsageProviderKind` and bumps **`USAGE_CONTRACT_VERSION` to 5**; the scan cache accepts **`grok`** entries.
> 
> Web and mobile usage UI gain **Grok Build** labels, chart bands, icons, and theme-aware colors; the time breakdown table’s empty-state **`colSpan`** follows `PROVIDER_ORDER` so a third provider column doesn’t break layout.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4c7a2762468c2db08e5498059fca12c17b761669. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add Grok Build as a provider in the usage view
> - Adds a `parseGrokLine` function in [usageTranscripts.ts](https://github.com/pingdotgg/t3code/pull/6051/files#diff-639e969b22f50fb1c28800a8caacb40faa709fb5b829f6c1e0a06a1be5fd5871) that parses Grok `updates.jsonl` session logs into per-model `UsageRecord` entries with cost tick-to-USD conversion and pro-rated cost attribution.
> - Extends the server-side usage scanner in [UsageService.ts](https://github.com/pingdotgg/t3code/pull/6051/files#diff-1d90d173860922c384f65961de4b51bba0176aadd7982190e0dcf87e663f1a70) to discover Grok sessions via `$GROK_HOME` or `~/.grok`, filtering to `updates.jsonl` files only.
> - Adds `"grok"` to `UsageProviderKind` and bumps the contract version from 4 to 5 in [usage.ts](https://github.com/pingdotgg/t3code/pull/6051/files#diff-9abc3595e224ffc63bc8ed068d884a2dad83ccb207f0b183e94259fb7f486a13).
> - Updates mobile and web UI provider registries to display Grok Build with a label, color, and icon in charts and tables.
> - Fixes the empty-state table colspan in [UsagePage.tsx](https://github.com/pingdotgg/t3code/pull/6051/files#diff-e076db2611ea0a07e1c8ec8571cc21103a2a4c7fc0feb865192fc12c8a18dead) to compute dynamically from `PROVIDER_ORDER.length` instead of a hardcoded value.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 4c7a276.</sup>
> 
>

### CODE REVIEW

File: apps/web/src/components/usage/usageProviders.ts [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
+import { ClaudeAI, GrokIcon, type Icon, OpenAI } from "../Icons";
 
 /**
- * Series and table order. The chart layers both providers from a shared zero
+ * Series and table order. The chart layers providers from a shared zero
  * baseline, so this only fixes the reading order of legends, tables and hover
  * rows; it does not decide which series sits above the other.
  */
-export const PROVIDER_ORDER: readonly UsageProviderKind[] = ["codex", "claude"];
+export const PROVIDER_ORDER: readonly UsageProviderKind[] = ["codex", "claude", "grok"];
```
Comment (EliasofIX) [inline apps/web/src/components/usage/usageProviders.ts]: **[nit]** Adding `"grok"` to `PROVIDER_ORDER` grows the daily breakdown table to 6 columns (Day + 3 providers + Total + Tokens), but `UsagePage` still hard-codes `colSpan={5}` on the empty-state row for that table. With no activity the “No activity in this window.” cell underspans the header row.

**Suggestion:** In `UsagePage.tsx`, set the day-table empty-state `colSpan` to `PROVIDER_ORDER.length + 3` (or `6` for now) so empty and populated headers stay aligned.

File: apps/server/src/usage/usageTranscripts.ts [hunk 2]
```diff
@@ -297,4 +310,164 @@ export function parseCodexLine(line: string, state: CodexScanState): UsageRecord
...
+  const results: UsageRecord[] = [];
+  for (const entry of modelEntries) {
+    const totals = grokTotalsToUsage(entry.totals);
+    if (totalTokens(totals) === 0) continue;
+
+    // Per-model ticks when present; otherwise a single-model turn may inherit
+    // the aggregate. Multi-model turns without per-model ticks stay unpriced
+    // at the provider-reported layer and fall through to the rate table.
+    let reportedCostUsd = grokCostTicksToUsd(entry.totals.costUsdTicks);
+    if (reportedCostUsd === null && modelEntries.length === 1) {
```
Comment (EliasofIX) [inline apps/server/src/usage/usageTranscripts.ts]: **[suggestion]** Per-model `costUsdTicks` inheritance only runs when `modelEntries.length === 1`. Multi-model turns that only carry top-level ticks intentionally leave `reportedCostUsd` null and fall through to LiteLLM; if those model slugs are missing from the rate table, the turn’s known provider cost is dropped and cost lands as unpriced/$0 while tokens remain.

**Suggestion:** If multi-model entries all lack ticks but the aggregate has them, either pro-rate the top-level ticks by token share, or fall back to a single aggregate record (as the no-`modelUsage` path already does) so provider-reported cost is not discarded. A focused test for that shape would lock the policy in.

File: apps/web/src/components/usage/usageProviders.ts [hunk 3]
```diff
@@ -1,23 +1,25 @@
...
   codex: "Codex",
+  grok: "Grok Build",
 };
 
-/** Claude's brand orange against a neutral white for Codex. */
+/** Claude's brand orange, neutral white for Codex, slate for Grok. */
 export const PROVIDER_COLOR: Record<UsageProviderKind, string> = {
   claude: "#d97757",
   codex: "#e6e6e6",
+  grok: "#a1a1aa",
```
Comment (EliasofIX) [inline apps/web/src/components/usage/usageProviders.ts]: **[nit]** Mobile flips Grok (and Codex) neutrals by color scheme so bars stay visible on light backgrounds; web keeps fixed dark-theme greys (`#a1a1aa` / `#e6e6e6`). On light web theme, Grok is acceptable mid-grey, but the surface pair is inconsistent and Codex remains nearly invisible on light—this change extends that pattern rather than matching mobile.

**Suggestion:** Prefer theme-aware colors on web (CSS variables or a light/dark pair) consistent with mobile’s `useProviderColors`, at least for the new Grok series.

File: apps/server/src/usage/usageTranscripts.ts [hunk 4]
```diff
@@ -297,4 +310,164 @@ export function parseCodexLine(line: string, state: CodexScanState): UsageRecord
...
+  }
+
+  // Older turns (or incomplete writes) may only carry the aggregate usage blob.
+  if (modelEntries.length === 0) {
+    if (totalTokens(grokTotalsToUsage(topLevel)) === 0) return [];
+    return [
+      {
+        provider: "grok",
+        timestampMs,
+        model: "grok",
```
Comment (EliasofIX) [inline apps/server/src/usage/usageTranscripts.ts:435]: **[nit]** The aggregate fallback when `modelUsage` is missing/empty (model labeled `"grok"`, top-level ticks) is implemented but not covered by tests. All `parseGrokLine` fixtures either supply `modelUsage` or zero it out without asserting the fallback record shape.

**Suggestion:** Add one case with no `modelUsage` (or empty object) that asserts `model === "grok"`, token totals from the aggregate blob, and top-level `costUsdTicks` conversion.

**Review by EliasofIX — COMMENTED** [review]
Submitting pending review so follow-up fix replies can land on the nits below.

File: apps/web/src/components/usage/usageProviders.ts (hunk 1 above)
Comment (EliasofIX) [reply apps/web/src/components/usage/usageProviders.ts]: Addressed in b1dfa9a48: day-table empty-state colSpan is now `PROVIDER_ORDER.length + 3` so it stays aligned when Grok is present.

File: apps/server/src/usage/usageTranscripts.ts (hunk 2 above)
Comment (EliasofIX) [reply apps/server/src/usage/usageTranscripts.ts]: Addressed in 0d56431a655d962959d1899e171ec784c5b3c311: multi-model turns that only carry top-level `costUsdTicks` now pro-rate those ticks by each model's totalTokens share, so provider-reported cost is not dropped. Covered by a focused unit test.

File: apps/web/src/components/usage/usageProviders.ts (hunk 3 above)
Comment (EliasofIX) [reply apps/web/src/components/usage/usageProviders.ts]: Addressed in 0d56431a655d962959d1899e171ec784c5b3c311: web Grok series color is now zinc-500 (`#71717a`) so it stays readable on both light and dark chart surfaces without a full theme-aware color map for this PR.

File: apps/server/src/usage/usageTranscripts.ts (hunk 4 above)
Comment (EliasofIX) [reply apps/server/src/usage/usageTranscripts.ts:435]: Addressed in 0d56431a655d962959d1899e171ec784c5b3c311: added a `modelUsage: null` fixture that asserts model `"grok"`, aggregate token totals, top-level cost tick conversion, and the fallback dedupe key.

File: apps/server/src/usage/usageTranscripts.ts [hunk 5]
```diff
@@ -297,4 +310,178 @@ export function parseCodexLine(line: string, state: CodexScanState): UsageRecord
...
+  // slugs). Single-model turns simply inherit the aggregate below.
+  const anyPerModelTicks = modelEntries.some((entry) => entry.totals.costUsdTicks !== null);
+  const topLevelCostUsd = grokCostTicksToUsd(topLevel.costUsdTicks);
+  let sharedTokenDenominator = 0;
+  if (!anyPerModelTicks && topLevelCostUsd !== null && modelEntries.length > 1) {
+    for (const entry of modelEntries) {
+      sharedTokenDenominator += totalTokens(grokTotalsToUsage(entry.totals));
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (EliasofIX) [reply apps/server/src/usage/usageTranscripts.ts:467]: Addressed in 16636a6bf74b9bc03961d0b09f397d14c909afba: `anyPerModelTicks` now ignores zero-token modelUsage rows (e.g. costUsdTicks: 0 empty siblings) so they no longer suppress aggregate pro-rating for billable models that omit per-model ticks. Added a regression test for the multi-model + empty-sibling case.

File: apps/server/src/usage/usageTranscripts.ts [hunk 6]
```diff
@@ -297,4 +310,184 @@ export function parseCodexLine(line: string, state: CodexScanState): UsageRecord
...
+  // When every per-model entry lacks ticks but the aggregate reports them,
+  // pro-rate the top-level ticks by each model's token share so provider cost
+  // is not discarded on multi-model turns (rate-table fallback can still miss
+  // slugs). Single-model turns simply inherit the aggregate below.
+  // Only count ticks on entries the emit loop would keep: zero-token siblings
+  // (often costUsdTicks: 0) must not suppress aggregate pro-rating for
+  // billable models that omit per-model ticks.
+  const anyPerModelTicks = modelEntries.some(
```
[the remark this answers is not in the document]
Comment (EliasofIX) [reply apps/server/src/usage/usageTranscripts.ts]: Addressed in 55b6f83ba77c2488956472c00ec328f111f10f52: models with their own costUsdTicks keep them; remaining aggregate cost (top-level minus those ticks, clamped at 0) is pro-rated across emitted models that omit ticks by token share among the unticked set. Zero-token rows still never count toward used ticks or emission.

[the remark this answers is not in the document]
Comment (EliasofIX) [reply apps/server/src/usage/usageTranscripts.ts]: Addressed in 55b6f83ba77c2488956472c00ec328f111f10f52: mixed per-model ticks now keep each model's own costUsdTicks, then pro-rate any leftover top-level aggregate ticks across emitted models that omitted ticks (by their token share). Zero-token siblings still do not suppress allocation.

**Comment by t3dotgg** [comment]
> [!NOTE]
> 🤖 **GPT-5.6 Sol responding on behalf of Theo**

I am closing this because #8358 has landed in `main` and covers the same Grok historical usage reporting. I checked the current implementation and did not find any remaining change from this PR that we still need to merge.

Thank you for putting this together. The closure is only because the work is now covered elsewhere, not because there was a problem with your contribution.
